### PR TITLE
feat(adc): impl basic ADC support

### PR DESCRIPTION
**Pull request by andelf**
- [x] ADC0
- [x] ADC1
- [x] interrupt enabling
- ~~[ ] DMA, blocked by missing dma mod~~

~~BTW, Haven't figured out how to impl a `hal::adc::OneShot`.~~

### CODE REVIEW

**Comment by andelf** [comment]
@Disasm PTAL

**Comment by andelf** [comment]
Usage: 

```rust
// ! use gd32vf103xx_hal::adc::{self, Adc, Temperature, Vrefint};

    let mut adc = Adc::adc0(dp.ADC0, &mut rcu);
    let config = adc::config::AdcConfig::default()
        .scan(adc::config::Scan::Disabled)
        .continuous(adc::config::Continuous::Single)
        .enable_inserted_channel(Default::default());
    adc.apply_config(config);

    let temp = Temperature::new().enable(&mut adc);
    adc.configure_inserted_channel(0, &temp, adc::config::SampleTime::Point_239_5);

    let mut adc = adc.enable();
    adc.calibrate();

    adc.enable_software_trigger();
    adc.wait_for_conversion();

    let raw_temp = adc.read_idata0();
    let temperature = (1.45 - (raw_temp as f32 * 3.3 / 4096.0)) * 1000.0 / 4.1 + 25.0;
    
    adc.clear_end_of_conversion_flag();
```

**Comment by Disasm** [comment]
Thank you @andelf! I'll try to review the PR this weekend.

File: src/adc.rs [hunk 1]
```diff
@@ -0,0 +1,807 @@
+//! Analog-to-digital converter
+//!
+use core::marker::PhantomData;
+
+// NOTE: embedded_hal's Channel is not suitable for this rank + channel style ADC.
```
Comment (Disasm) [inline src/adc.rs:5]: Could you elaborate on this? I believe that GD32V ADC should be "the same" as the one used in STM32F103, and the stm32f1xx-hal ADC driver implements embedded-hal traits.

Comment (andelf) [reply src/adc.rs:5]: They are not the same.

GD32VF103's channel number is actually the `rank` number when configuring, not the `Channel::channel()` number.

`Channel::channel()` in `embedded_hal` is a method bonded to type, not the self-object. So one can not save the real channel number in a wrapped "Channel" type.

Comment (andelf) [reply src/adc.rs:5]: The `OneShot` abstraction is not suitable for GD32VF103. So I left it unimplemented.

Comment (elionne) [reply src/adc.rs:5]: What do you mean by rank ? I mean there is 18 multiplexed channels like on stm32f103.

From my understanding, you can implement The OneShot mode as describe in documentation in section "Single conversion mode" with the sequence configured with only the desired channel.

Comment (andelf) [reply src/adc.rs:5]: `rank` is the parameter name defined in C SDK.
`Single conversion mode` might help, will try to implement it. 😁

**Comment by andelf** [comment]
OneShot now works. Usage:

```rust

    let mut adc = Adc::adc0(dp.ADC0, &mut rcu);
    let config = adc::config::AdcConfig::default()
        .enable_regular_channel(Default::default());
    adc.apply_config(config);

    let mut vrefint = Vrefint::new().enable(&mut adc);
    let mut temp = Temperature::new().enable(&mut adc);

    let mut adc = adc.enable();
    adc.calibrate();

    let raw_temp: u16 = adc.read(&mut temp).unwrap();
    let temperature = (1.45 - (raw_temp as f32 * 3.3 / 4096.0)) * 1000.0 / 4.1 + 25.0;
    let raw_vref: u16 = adc.read(&mut vrefint).unwrap();
    let vref_value = raw_vref as f32 * 3.3 / 4096.0;
```

**Comment by andelf** [comment]
PTAL @Disasm @elionne

File: src/adc.rs [hunk 2]
```diff
@@ -0,0 +1,858 @@
...
+        /// ADC prescaler select CK_APB2/16
+        Apb2_div_16 = 7,
+    }
+
+    /// Resolution to sample at
+    #[derive(Debug, Clone, Copy)]
+    #[repr(u8)]
+    pub enum Resolution {
+        /// 12-bit ADC resolution
+        Twelve = 0,
```
Comment (elionne) [inline src/adc.rs:84]: To be coherent with other enums, maybe something like `Res_12, … Res_6` would be preferable.

File: src/adc.rs [hunk 3]
```diff
@@ -0,0 +1,858 @@
...
+
+    impl RegularChannelGroupConfig {
+        /// change the external_trigger field
+        pub fn external_trigger(mut self, external_trigger: RegularExternalTrigger) -> Self {
+            self.external_trigger = external_trigger;
+            self
+        }
+    }
+
+    impl Default for RegularChannelGroupConfig {
```
Comment (elionne) [inline src/adc.rs:201]: Why you do not just implement `new()` for `RegularChannelGroupConfig` ?

File: src/adc.rs [hunk 4]
```diff
@@ -0,0 +1,858 @@
...
+            self
+        }
+        /// change the default_sample_time field
+        pub fn default_sample_time(mut self, default_sample_time: SampleTime) -> Self {
+            self.default_sample_time = default_sample_time;
+            self
+        }
+    }
+
+    impl Default for AdcConfig {
```
Comment (elionne) [inline src/adc.rs:309]: Again, why not a simple `new()`. ?

File: src/adc.rs [hunk 5]
```diff
@@ -0,0 +1,858 @@
...
+
+    /// Enables and disables scan mode
+    pub fn set_scan(&mut self, scan: config::Scan) {
+        self.config.scan = scan;
+    }
+
+    /// Enables and disables continuous mode
+    pub fn set_continuous(&mut self, continuous: config::Continuous) {
+        self.config.continuous = continuous;
+    }
```
Comment (elionne) [inline src/adc.rs:429]: Does this methods are necessary ? Because they are already implemented in AdcConfig ?

File: src/adc.rs [hunk 6]
```diff
@@ -0,0 +1,858 @@
...
+        unsafe {
+            self.rb
+                .ovsampctl
+                .modify(|_, w| w.dres().bits(resolution as u8));
+        }
+    }
+}
+
+impl Adc<ADC1, Disabled> {
+    fn pre_configure(&mut self, _resolution: config::Resolution) {}
```
Comment (elionne) [inline src/adc.rs:445]: Why there is no `ovsampctl` register in ADC1 ?

Comment (andelf) [reply src/adc.rs:445]: https://docs.rs/gd32vf103-pac/0.4.0/gd32vf103_pac/adc1/struct.RegisterBlock.html

File: src/adc.rs (hunk 2 above)
Comment (andelf) [reply src/adc.rs:84]: https://docs.rs/stm32f4xx-hal/0.8.3/stm32f4xx_hal/adc/config/enum.Resolution.html

File: src/adc.rs (hunk 5 above)
Comment (andelf) [reply src/adc.rs:429]: https://docs.rs/stm32f4xx-hal/0.8.3/stm32f4xx_hal/adc/struct.Adc.html#method.set_clock

**Review by sajattack — APPROVED** [review]

**Comment by dkhayes117** [comment]
Closed due to another ADC PR #38 being accepted.

**Pull request closed without merging**